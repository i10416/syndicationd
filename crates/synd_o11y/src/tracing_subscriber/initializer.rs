use crate::{
    opentelemetry::init_propagation,
    opentelemetry_layer,
    tracing_subscriber::{audit, otel_metrics::metrics_event_filter},
    OpenTelemetryGuard,
};

const LOG_DIRECTIVE: &str = "SYND_LOG";
const DEFAULT_LOG_DIRECTIVE: &str = "info";

pub struct TracingInitializer {
    app_name: Option<&'static str>,
    app_version: Option<&'static str>,
    otlp_endpoint: Option<String>,
    trace_sampler_ratio: f64,
    enable_ansi: bool,
    show_code_location: bool,
    show_target: bool,
}

impl Default for TracingInitializer {
    fn default() -> Self {
        Self {
            app_name: None,
            app_version: None,
            otlp_endpoint: None,
            trace_sampler_ratio: 1.,
            enable_ansi: true,
            show_code_location: false,
            show_target: true,
        }
    }
}

impl TracingInitializer {
    /// Initialize tracing Subscriber with layers
    pub fn init(self) -> Option<OpenTelemetryGuard> {
        use tracing_subscriber::{
            filter::EnvFilter, fmt, layer::SubscriberExt, util::SubscriberInitExt as _, Layer as _,
            Registry,
        };

        let Self {
            app_name: Some(app_name),
            app_version: Some(app_version),
            otlp_endpoint,
            trace_sampler_ratio,
            enable_ansi,
            show_code_location,
            show_target,
        } = self
        else {
            panic!()
        };

        let (otel_layer, otel_guard) =
            match otlp_endpoint
                .as_deref()
                .filter(|s| !s.is_empty())
                .map(|endpoint| {
                    opentelemetry_layer(endpoint, app_name, app_version, trace_sampler_ratio)
                }) {
                Some((otel_layer, otel_guard)) => (Some(otel_layer), Some(otel_guard)),
                _ => (None, None),
            };

        Registry::default()
            .with(
                fmt::Layer::new()
                    .with_ansi(enable_ansi)
                    .with_timer(fmt::time::ChronoLocal::rfc_3339())
                    .with_file(show_code_location)
                    .with_line_number(show_code_location)
                    .with_target(show_target)
                    .with_filter(metrics_event_filter())
                    .and_then(otel_layer)
                    .with_filter(
                        EnvFilter::try_from_env(LOG_DIRECTIVE)
                            .or_else(|_| EnvFilter::try_new(DEFAULT_LOG_DIRECTIVE))
                            .unwrap()
                            .add_directive(audit::Audit::directive()),
                    ),
            )
            .with(audit::layer())
            .init();

        // Set text map propagator globally
        init_propagation();

        otel_guard
    }
}

impl TracingInitializer {
    #[must_use]
    pub fn app_name(self, app_name: &'static str) -> Self {
        Self {
            app_name: Some(app_name),
            ..self
        }
    }

    #[must_use]
    pub fn app_version(self, app_version: &'static str) -> Self {
        Self {
            app_version: Some(app_version),
            ..self
        }
    }

    #[must_use]
    pub fn otlp_endpoint(self, otlp_endpoint: Option<String>) -> Self {
        Self {
            otlp_endpoint,
            ..self
        }
    }

    #[must_use]
    pub fn trace_sampler_ratio(self, trace_sampler_ratio: f64) -> Self {
        Self {
            trace_sampler_ratio,
            ..self
        }
    }

    #[must_use]
    pub fn enable_ansi(self, enable_ansi: bool) -> Self {
        Self {
            enable_ansi,
            ..self
        }
    }

    #[must_use]
    pub fn show_code_location(self, show_code_location: bool) -> Self {
        Self {
            show_code_location,
            ..self
        }
    }

    #[must_use]
    pub fn show_target(self, show_target: bool) -> Self {
        Self {
            show_target,
            ..self
        }
    }
}
