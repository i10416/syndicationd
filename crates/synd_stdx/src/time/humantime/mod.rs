mod parse;
pub use parse::{parse_duration, DurationError};
