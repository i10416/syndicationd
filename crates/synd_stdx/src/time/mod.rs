#[cfg(feature = "humantime")]
pub mod humantime;
