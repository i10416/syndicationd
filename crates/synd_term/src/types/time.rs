pub type Time = synd_feed::types::Time;

pub trait TimeExt {
    fn local_ymd(&self) -> String;
}

impl TimeExt for Time {
    fn local_ymd(&self) -> String {
        self.naive_local().format("%Y/%m/%d").to_string()
    }
}
