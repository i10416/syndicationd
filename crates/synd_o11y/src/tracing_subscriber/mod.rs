use std::{borrow::Cow, time::Duration};

use opentelemetry_sdk::{logs, trace};
use tracing::Subscriber;
use tracing_subscriber::{Layer, registry::LookupSpan};

use crate::{OpenTelemetryGuard, tracing_subscriber::otel_metrics::metrics_event_filter};

pub mod audit;
pub mod initializer;
pub mod otel_log;
pub mod otel_metrics;
pub mod otel_trace;

/// Return a composed layer that exports traces, metrics, and logs of opentelemetry.
/// The exported telemetry will be accompanied by a `Resource`.
pub fn opentelemetry_layer<S>(
    endpoint: impl Into<String>,
    service_name: impl Into<Cow<'static, str>>,
    service_version: impl Into<Cow<'static, str>>,
    trace_sampler_ratio: f64,
) -> (impl Layer<S>, OpenTelemetryGuard)
where
    S: Subscriber + for<'span> LookupSpan<'span>,
{
    let endpoint = endpoint.into();
    let resource = crate::opentelemetry::resource(service_name, service_version);

    let (trace_layer, tracer_provider) = {
        let trace_batch_config = trace::BatchConfigBuilder::default().build();
        otel_trace::layer(
            endpoint.clone(),
            resource.clone(),
            trace_sampler_ratio,
            trace_batch_config,
        )
    };

    let (metrics_layer, meter_provider) = {
        let metrics_reader_interval = Duration::from_secs(60);
        otel_metrics::layer(endpoint.clone(), resource.clone(), metrics_reader_interval)
    };

    let (log_layer, logger_provider) = {
        let log_batch_config = logs::BatchConfigBuilder::default().build();
        otel_log::layer(endpoint.clone(), resource.clone(), log_batch_config)
    };

    // Since metrics events are handled by the metrics layer and are not needed as logs, set a filter for them.
    let log_layer = log_layer.with_filter(metrics_event_filter());

    let guard = OpenTelemetryGuard {
        tracer_provider,
        meter_provider,
        logger_provider,
    };

    (
        trace_layer.and_then(metrics_layer).and_then(log_layer),
        guard,
    )
}
