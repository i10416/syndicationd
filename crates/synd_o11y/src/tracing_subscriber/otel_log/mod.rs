use opentelemetry_appender_tracing::layer::OpenTelemetryTracingBridge;
use opentelemetry_otlp::WithExportConfig as _;
use opentelemetry_sdk::{runtime, Resource};
use tracing::Subscriber;
use tracing_subscriber::{registry::LookupSpan, Layer};

pub fn layer<S>(endpoint: impl Into<String>, resource: Resource) -> impl Layer<S>
where
    S: Subscriber + for<'span> LookupSpan<'span>,
{
    opentelemetry_otlp::new_pipeline()
        .logging()
        .with_log_config(opentelemetry_sdk::logs::Config::default().with_resource(resource))
        .with_exporter(
            opentelemetry_otlp::new_exporter()
                .tonic()
                .with_endpoint(endpoint),
        )
        .install_batch(runtime::Tokio)
        .unwrap();

    let provider = opentelemetry::global::logger_provider();

    OpenTelemetryTracingBridge::new(&provider)
}
