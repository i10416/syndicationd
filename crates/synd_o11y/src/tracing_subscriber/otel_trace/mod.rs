use opentelemetry::{global, trace::TracerProvider};
use opentelemetry_otlp::WithExportConfig;
use opentelemetry_sdk::{
    runtime,
    trace::{BatchConfig, Sampler, Tracer},
    Resource,
};
use tracing::Subscriber;
use tracing_opentelemetry::OpenTelemetryLayer;
use tracing_subscriber::{registry::LookupSpan, Layer};

pub fn layer<S>(
    endpoint: impl Into<String>,
    resource: Resource,
    sampler_ratio: f64,
    batch_config: BatchConfig,
) -> impl Layer<S>
where
    S: Subscriber + for<'span> LookupSpan<'span>,
{
    OpenTelemetryLayer::new(init_tracer(endpoint, resource, sampler_ratio, batch_config))
}

#[expect(clippy::needless_pass_by_value)]
fn init_tracer(
    endpoint: impl Into<String>,
    resource: Resource,
    sampler_ratio: f64,
    // TODO: how to use BatchConfig after 0.27 ?
    _batch_config: BatchConfig,
) -> Tracer {
    let exporter = opentelemetry_otlp::SpanExporter::builder()
        .with_tonic()
        .with_endpoint(endpoint)
        .build()
        .unwrap();

    let provider = opentelemetry_sdk::trace::TracerProvider::builder()
        .with_resource(resource)
        .with_sampler(Sampler::ParentBased(Box::new(Sampler::TraceIdRatioBased(
            sampler_ratio,
        ))))
        .with_batch_exporter(exporter, runtime::Tokio)
        .build();

    // > It would now be the responsibility of users to set it by calling global::set_tracer_provider(tracer_provider.clone());
    //  https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-otlp/CHANGELOG.md#v0170
    global::set_tracer_provider(provider.clone());

    provider.tracer("tracing-opentelemetry")
}

#[cfg(test)]
mod tests {
    use std::{net::SocketAddr, time::Duration};

    use opentelemetry::KeyValue;
    use opentelemetry_proto::tonic::{
        collector::trace::v1::{
            trace_service_server::{TraceService, TraceServiceServer},
            ExportTraceServiceRequest, ExportTraceServiceResponse,
        },
        trace::v1::{span::SpanKind, status::StatusCode, Status},
    };
    use opentelemetry_sdk::trace::BatchConfigBuilder;
    use tokio::sync::mpsc::UnboundedSender;
    use tonic::transport::Server;
    use tracing::dispatcher;
    use tracing_subscriber::{layer::SubscriberExt, Registry};

    use super::*;

    type Request = tonic::Request<ExportTraceServiceRequest>;

    struct DumpTraces {
        tx: UnboundedSender<Request>,
    }

    #[tonic::async_trait]
    impl TraceService for DumpTraces {
        async fn export(
            &self,
            request: tonic::Request<ExportTraceServiceRequest>,
        ) -> Result<tonic::Response<ExportTraceServiceResponse>, tonic::Status> {
            self.tx.send(request).unwrap();

            Ok(tonic::Response::new(ExportTraceServiceResponse {
                partial_success: None, // means success
            }))
        }
    }

    #[tracing::instrument(fields(
        otel.name = "f1_custom",
        otel.kind = "Client",
    ) )]
    fn f1() {
        f2();
    }
    #[tracing::instrument(fields(
        otel.name = "f2_custom",
        otel.kind = "Server",
    ))]
    fn f2() {
        tracing::error!("error_f2");
    }

    #[tokio::test(flavor = "multi_thread")]
    async fn layer_test() {
        let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
        let dump = TraceServiceServer::new(DumpTraces { tx });
        let addr: SocketAddr = ([127, 0, 0, 1], 48100).into();
        let server = Server::builder().add_service(dump).serve(addr);
        let _server = tokio::task::spawn(server);
        let resource = resource();
        let config = BatchConfigBuilder::default()
            // The default interval is 5 seconds, which slows down the test
            .with_scheduled_delay(Duration::from_millis(10))
            .build();
        let layer = layer("https://localhost:48100", resource.clone(), 1.0, config);
        let subscriber = Registry::default().with(layer);
        let dispatcher = tracing::Dispatch::new(subscriber);

        dispatcher::with_default(&dispatcher, || {
            f1();
        });

        let req = rx.recv().await.unwrap().into_inner();
        assert_eq!(req.resource_spans.len(), 1);

        let resource_span = req.resource_spans[0].clone();
        insta::with_settings!({
            description => "trace resource",
        }, {
            insta::assert_yaml_snapshot!("layer_test_trace_resource", req.resource_spans[0].resource);
        });

        let [f2_span, f1_span] = resource_span.scope_spans[0].spans.as_slice() else {
            panic!()
        };

        assert_eq!(&f2_span.name, "f2_custom");
        assert_eq!(&f1_span.name, "f1_custom");
        assert_eq!(f2_span.parent_span_id, f1_span.span_id);
        assert_eq!(f2_span.trace_id, f1_span.trace_id);
        assert_eq!(f2_span.kind, SpanKind::Server as i32);
        assert_eq!(f1_span.kind, SpanKind::Client as i32);
        assert_eq!(
            f2_span.status,
            Some(Status {
                message: String::new(),
                code: StatusCode::Error as i32,
            })
        );
        assert_eq!(
            f1_span.status,
            Some(Status {
                message: String::new(),
                code: StatusCode::Unset as i32,
            })
        );
        assert_eq!(f2_span.events.len(), 1);
        assert_eq!(f2_span.attributes.len(), 7);
    }

    fn resource() -> Resource {
        Resource::new([KeyValue::new("service.name", "test")])
    }
}
