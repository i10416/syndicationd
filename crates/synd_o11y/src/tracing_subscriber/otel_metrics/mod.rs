use std::time::Duration;

use opentelemetry_sdk::{metrics::MeterProvider, runtime, Resource};
use tracing::{Metadata, Subscriber};
use tracing_opentelemetry::MetricsLayer;
use tracing_subscriber::{filter::filter_fn, layer::Filter, registry::LookupSpan, Layer};

pub mod macros;

pub const METRICS_EVENT_TARGET: &str = "metrics";

pub fn metrics_event_filter<S: Subscriber>() -> impl Filter<S> {
    filter_fn(|metadata: &Metadata<'_>| metadata.target() != METRICS_EVENT_TARGET)
}

pub fn layer<S>(resource: Resource) -> impl Layer<S>
where
    S: Subscriber + for<'span> LookupSpan<'span>,
{
    MetricsLayer::new(init_meter_provider(resource))
}

fn init_meter_provider(resource: Resource) -> MeterProvider {
    let provider = opentelemetry_otlp::new_pipeline()
        .metrics(runtime::Tokio)
        .with_resource(resource)
        .with_exporter(opentelemetry_otlp::new_exporter().tonic())
        .with_period(Duration::from_secs(60))
        .build()
        .unwrap();

    opentelemetry::global::set_meter_provider(provider.clone());

    provider
}
