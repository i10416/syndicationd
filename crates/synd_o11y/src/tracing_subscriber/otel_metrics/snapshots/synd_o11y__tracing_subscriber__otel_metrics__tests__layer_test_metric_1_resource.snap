---
source: crates/synd_o11y/src/tracing_subscriber/otel_metrics/mod.rs
description: " metric 1 resource"
expression: metric1.resource
---
attributes:
  - key: service.name
    value:
      stringValue: test
  - key: telemetry.sdk.language
    value:
      stringValue: rust
  - key: telemetry.sdk.name
    value:
      stringValue: opentelemetry
  - key: telemetry.sdk.version
    value:
      stringValue: 0.29.0
droppedAttributesCount: 0
